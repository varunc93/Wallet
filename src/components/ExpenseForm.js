import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) { //This is done to get access to props
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : "",
            amount: props.expense ? (props.expense.amount/100).toString() : "",
            text: props.expense ? props.expense.text : "",
            createdAt: props.expense ? moment(props.expense.createdAt): moment(),
            focused: false,
            error:""
        }
    };

    handleDescription = (event) => {
        const description = event.target.value;
        this.setState(() => ({description}));
    };

    //OR
    // handleDescription = (event) => {
    //     event.persist;
    //     this.setState(() => ({description: event.target.value}));
    // }

    handleAmount = (event) => {
        const amount = event.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) //We use regular expressions to check validation. ^-ensures the user starts with a number
        //\d-ensures it's a number, *- to allow multiple digits, ()-allows optional parameters,
        //?-ensures the parameter is only allowed once, \.-allows one decimal digit, {0,2}-allows upto 2 characters
        //$ ensures we end the input with the regular expression mentioned
        //match() takes in regular expression, /<place regular expression here>/
        // !amount allows user to clear the data in the amount box
        this.setState(() => ({amount}));
    };

    handleText = (event) => {
        const text = event.target.value;
        this.setState(() => ({text}));
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
          this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({focused}));
    };

    onSubmit = (event) => {
        event.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState((error) => true);
            if(!this.state.description){
               this.setState(() => ({error: "Error: Please enter description"}))
            }
            else{
                this.setState(() => ({error: "Error: Please enter amount"}))
            }
        }
        else {
            this.setState(() => ({error: ""}))
            this.props.onSubmit({ //Props to be passed to Add and Edit Expense Pages
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //Since amount is stored as string
                text: this.state.text,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };

    //Button is placed inside a div to prevent it from being a direct descendant.
    render () {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input className="text-input" type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.handleDescription}/>
                <input className="text-input" type="text" placeholder="Amount" value= {this.state.amount} onChange={this.handleAmount}/>
                <textarea className="textarea" placeholder="Add a note..." value={this.state.text} onChange={this.handleText}></textarea>
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    };

}
