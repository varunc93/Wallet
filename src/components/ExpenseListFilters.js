import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        focused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (focused) => { //focused is the new value
        this.setState(() => ({focused})); //Same as this.setState(() => ({focused: focused}));
    };

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value);
    };

    onSortChange = (event) => {
        if(event.target.value === "date"){
            this.props.sortByDate();
        }
        else if(event.target.value === "amount"){
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    startDateId="start"
                    endDateId="end"
                />
            </div>
        )
    }
};

//Basic Template
// const mapStateToProps = (state) => {
//     return {

//     }
// }

const mapStateToProps = (state) => { //Access to whole and we need to filter out things we want and don't want and pass it through connect.
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);