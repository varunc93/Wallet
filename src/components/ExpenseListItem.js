import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

// const ExpenseListItem = (props) => (
//     <div>
//         <Link to={`/edit/${props.id}`}>
//             <h3>{props.description}</h3>
//         </Link>
//         <p>{props.amount} - {props.createdAt}</p>
//     </div>
// );

//Same as
const ExpenseListItem = ({id, description, amount, createdAt}) => {
    return (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{numeral(amount/100).format('$0,0.00')}
            -
            {moment(createdAt).format("MMMM Do, YYYY")}</p>
    </div>
)};

//We need to use connect to be able to access dispatch
export default ExpenseListItem;