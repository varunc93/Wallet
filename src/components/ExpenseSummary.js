import React from 'react';
import { connect } from 'react-redux';
import {Link} from 'react-router-dom';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? "expense" : "expenses";
    const formattedExpense = numeral(expensesTotal/100).format("$0,0.00");
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page_header__title">Viewing a total of <span>{formattedExpense}</span> for <span>{expenseCount}</span> {expenseWord}. </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add New Expense</Link>
                </div>
            </div>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }
};

export default connect(mapStateToProps)(ExpenseSummary);