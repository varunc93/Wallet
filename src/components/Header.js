import React from 'react';
import {NavLink} from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({startLogout}) => (
    <header>
        <h1>Wallet</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Homepage</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <button onClick={startLogout}>Logout</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);

