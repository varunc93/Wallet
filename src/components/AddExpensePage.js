import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        this.props.history.push("/"); //push takes in a path to reroute to a stored path
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit = {this.onSubmit} //Props received from Expense Form
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({addExpense: (expense) => dispatch(addExpense(expense))}) //To make it easier for testing purposes

export default connect(undefined, mapDispatchToProps)(AddExpensePage);