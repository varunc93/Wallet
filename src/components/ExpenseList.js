import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = (props) => ( //Export required for testing to isolate ExpenseList and use custom expenses
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>There are no expenses to show!</span>
                    </div>
                ):
                (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />;
                }))
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
}

//Same as
// const mapStateToProps = (state) => ({
//     expenses: state.expenses,
//     filters: state.filters
// });


export default connect(mapStateToProps)(ExpenseList); //Connect returns a function