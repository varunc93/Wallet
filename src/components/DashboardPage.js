import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const DashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;