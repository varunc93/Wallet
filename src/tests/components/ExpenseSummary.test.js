import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import toJSON from 'enzyme-to-json';
import expense from "../fixtures/expenses";

test("ExpenseSummary should render 1 expense correctly", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test("ExpenseSummary should render multiple expenses correctly", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={4} expensesTotal={10392389} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
});


