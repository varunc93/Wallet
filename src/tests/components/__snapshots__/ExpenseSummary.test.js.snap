// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseSummary should render 1 expense correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page_header__title"
    >
      Viewing a total of 
      <span>
        $2.35
      </span>
       for 
      <span>
        1
      </span>
       
      expense
      . 
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add New Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`ExpenseSummary should render multiple expenses correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page_header__title"
    >
      Viewing a total of 
      <span>
        $103,923.89
      </span>
       for 
      <span>
        4
      </span>
       
      expenses
      . 
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add New Expense
      </Link>
    </div>
  </div>
</div>
`;
